MagneticField/ParametrizedEngine/plugins/AutoParametrizedMagneticFieldProducer.cc

```diff
@@ -0,0 +1,109 @@
+/** \class AutoParametrizedMagneticFieldProducer
+ *
+ *   Description: Producer for parametrized Magnetics Fields, with value scaled depending on current.
+ *
+ */
+
+#include "FWCore/Framework/interface/ESProducer.h"
+
+#include "MagneticField/Engine/interface/MagneticField.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "MagneticField/ParametrizedEngine/interface/ParametrizedMagneticFieldFactory.h"
+
+#include "MagneticField/Records/interface/IdealMagneticFieldRecord.h"
+#include "CondFormats/RunInfo/interface/RunInfo.h"
+#include "CondFormats/DataRecord/interface/RunSummaryRcd.h"
+
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include <FWCore/MessageLogger/interface/MessageLogger.h>
+
+#include <string>
+#include <iostream>
+
+using namespace std;
+using namespace edm;
+
+namespace magneticfield{
+  class AutoParametrizedMagneticFieldProducer : public edm::ESProducer
+  {
+  public:
+    AutoParametrizedMagneticFieldProducer(const edm::ParameterSet&);
+    ~AutoParametrizedMagneticFieldProducer(){}
+    
+    std::auto_ptr<MagneticField> produce(const IdealMagneticFieldRecord&);
+
+    int closerNominaCurrent(float current);
+    edm::ParameterSet pset;
+    std::vector<int> nominalCurrents;
+    //  std::vector<std::string> nominalLabels;
+  };
+}
+
+using namespace magneticfield;
+
+AutoParametrizedMagneticFieldProducer::AutoParametrizedMagneticFieldProducer(const edm::ParameterSet& iConfig) : pset(iConfig) {
+  setWhatProduced(this, pset.getUntrackedParameter<std::string>("label",""));
+  nominalCurrents={-1, 0,9558,14416,16819,18268,19262};
+  //  nominalLabels  ={"3.8T","0T","2T", "3T", "3.5T", "3.8T", "4T"};
+}
+
+std::auto_ptr<MagneticField>
+AutoParametrizedMagneticFieldProducer::produce(const IdealMagneticFieldRecord& iRecord)
+{
+  string version = pset.getParameter<string>("version");
+
+  // Get value of the current from condition DB
+  float current = pset.getParameter<int>("valueOverride");
+  string message;
+  if (current < 0) {
+    ESHandle<RunInfo> rInfo;
+    iRecord.getRecord<RunInfoRcd>().get(rInfo);
+    current = rInfo->m_avg_current;
+    message = " (from RunInfo DB)";
+  } else {
+    message = " (from valueOverride card)";
+  }
+  float cnc= closerNominaCurrent(current);
+
+  edm::LogInfo("MagneticField|AutoParametrizedMagneticField") << "Current: " << current << message << "; using map for: " << cnc;
+
+  vector<double> parameters;
+  
+  if (cnc==0) {
+    version = "Uniform";
+    parameters.push_back(0);
+  }
+
+  else if (version=="Parabolic"){
+    parameters.push_back(3.8114);       //c1
+    parameters.push_back(-3.94991e-06); //b0
+    parameters.push_back(7.53701e-06);  //b1
+    parameters.push_back(2.43878e-11);   //a
+    if (cnc!=18268){ // Linear scaling for B!= 3.8T; note that just c1, b0 and b1 have to be scaled to get linear scaling
```

Comment:
To very good approxymation,  inside the solenoid, it scales with 
current; please note that I don't scale with the current itself but with 
the closest nominal working point. Nominal WPs are 2, 3, 3.5 and 3.8T; 
maps are available only for these, and likewise I made this fast 
parametrization stick to the same.

At some point we could re-fit the 2, 3, 3.5T parameters, but that does 
not seem to be so fundamental, if we have short runs at B other than 
3.8T we can live with the decent approx of linear scaling.

Cheers
	Nicola

On 08-May-15 3:57, Slava Krutelyov wrote:
> In
> MagneticField/ParametrizedEngine/plugins/AutoParametrizedMagneticFieldProducer.cc
> <https://github.com/cms-sw/cmssw/pull/8992#discussion_r29912448>:
>
>> +
>> +  edm::LogInfo("MagneticField|AutoParametrizedMagneticField") << "Current: " << current << message << "; using map for: " << cnc;
>> +
>> +  vector<double> parameters;
>> +
>> +  if (cnc==0) {
>> +    version = "Uniform";
>> +    parameters.push_back(0);
>> +  }
>> +
>> +  else if (version=="Parabolic"){
>> +    parameters.push_back(3.8114);       //c1
>> +    parameters.push_back(-3.94991e-06); //b0
>> +    parameters.push_back(7.53701e-06);  //b1
>> +    parameters.push_back(2.43878e-11);   //a
>> +    if (cnc!=18268){ // Linear scaling for B!= 3.8T; note that just c1, b0 and b1 have to be scaled to get linear scaling
>
> Does this actually scale, especially for low currents?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/8992/files#r29912448>.
>